### PR TITLE
.Net: Added example with telemetry in Filters

### PR DESCRIPTION
**Pull request by dmytrostruk**
### Motivation and Context

Related: https://github.com/microsoft/semantic-kernel/issues/5436

Kernel and connectors have out-of-the-box telemetry to capture key information, which is available during requests.
In most cases this telemetry should be enough to understand how the application behaves.
This example contains the same telemetry recreated using Filters.
This should allow to extend existing telemetry if needed with additional information and have the same set of logging messages for custom connectors.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

File: dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            await next(context);
+
+            TimeSpan duration = new((long)((Stopwatch.GetTimestamp() - startingTimestamp) * (10_000_000.0 / Stopwatch.Frequency)));
+
+            logger.LogInformation("Function {FunctionName} succeeded.", context.Function.Name);
+            logger.LogTrace("Function result: {Result}", context.Result.ToString());
+
+            if (logger.IsEnabled(LogLevel.Information))
+            {
+                logger.LogInformation("Function completed. Duration: {Duration}s", duration.TotalSeconds);
```
Comment (westey-m) [inline dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:98]: I realise this is only an example, but milliseconds is probably more appropriate, since most function calls should not take seconds to complete, and it'll be easier to not work with fractions.

File: dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+
+    /// <summary>
+    /// Filter which logs an information available during function invocation such as:
+    /// Function name, arguments, execution settings, result, duration, token usage.
+    /// </summary>
+    private sealed class FunctionInvocationLoggingFilter(ILogger logger) : IFunctionInvocationFilter
+    {
+        public async Task OnFunctionInvocationAsync(FunctionInvocationContext context, Func<FunctionInvocationContext, Task> next)
+        {
+            long startingTimestamp = Stopwatch.GetTimestamp();
```
Comment (westey-m) [inline dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:75]: Why not use `var stopwatch = StopWatch.StartNew();` with `stopWatch.ElapsedMilliseconds`.
Seems simpler than having to do the calculation on line 91.

File: dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+            logger.LogInformation("Function {FunctionName} invoking.", context.Function.Name);
+
+            if (context.Arguments.Count > 0)
+            {
+                logger.LogTrace("Function arguments: {Arguments}", context.Arguments);
```
Comment (westey-m) [inline dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:81]: Does arguments also need to be serialized?

File: dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    /// Filter which logs an information available during function invocation such as:
+    /// Function name, arguments, execution settings, result, duration, token usage.
+    /// </summary>
+    private sealed class FunctionInvocationLoggingFilter(ILogger logger) : IFunctionInvocationFilter
+    {
+        public async Task OnFunctionInvocationAsync(FunctionInvocationContext context, Func<FunctionInvocationContext, Task> next)
+        {
+            long startingTimestamp = Stopwatch.GetTimestamp();
+
+            logger.LogInformation("Function {FunctionName} invoking.", context.Function.Name);
```
Comment (westey-m) [inline dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:77]: I wonder how easy it would be for someone to correlate the different messages together here.  Would make sense to use something like a scope here with an id for this function invocation?
https://learn.microsoft.com/en-us/dotnet/core/extensions/logging?tabs=command-line#log-scopes

**Review by westey-m — APPROVED** [review]

File: dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs (hunk 1 above)
Comment (dmytrostruk) [reply dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:98]: All telemetry in this example is recreated from already existing telemetry in Semantic Kernel, based on requirement from original issue. Here is a log message from SK:
https://github.com/microsoft/semantic-kernel/blob/4ebd11b05d78e2f1800bbeae8b3c795c1d66a1ca/dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionLogMessages.cs#L116-L118

File: dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs (hunk 2 above)
Comment (dmytrostruk) [reply dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:75]: Same here, I decided to re-use some code which already exists :)
https://github.com/microsoft/semantic-kernel/blob/4ebd11b05d78e2f1800bbeae8b3c795c1d66a1ca/dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunction.cs#L223-L225

File: dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs (hunk 3 above)
Comment (dmytrostruk) [reply dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:81]: By default, `IDictionary` is handled by logger correctly, that's why I don't see any issues in output, I also tested this behavior in empty console application:
```csharp
var dictionary = new Dictionary<string, string> { ["key"] = "value" };

logger.LogInformation("Test: {Dictionary}", dictionary);
```

Output:
```
info: ConsoleApp1.Program[0]
      Test: [key, value]
```

But I see in SK telemetry we actually serialize it, so maybe it will be better to do that in example as well, I will make the update:
https://github.com/microsoft/semantic-kernel/blob/9ac1152a1f5cb402e623ca4bea0b3f6019a7d47f/dotnet/src/SemanticKernel.Abstractions/Functions/KernelFunctionLogMessages.cs#L45

File: dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+            {
+                logger.LogTrace("Function arguments: {Arguments}", context.Arguments);
+            }
+
+            if (logger.IsEnabled(LogLevel.Information) && context.Arguments.ExecutionSettings is not null)
+            {
+                logger.LogInformation("Execution settings: {Settings}", JsonSerializer.Serialize(context.Arguments.ExecutionSettings));
+            }
+
+            await next(context);
```
[the remark this answers is not in the document]
Comment (dmytrostruk) [reply dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs]: Good catch, thank you! Added.

File: dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs (hunk 3 above)
Comment (dmytrostruk) [reply dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs]: Updated with arguments serialization.

File: dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs (hunk 4 above)
Comment (dmytrostruk) [reply dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:112]: Yes, I was thinking about that too, but it's important to understand that scoping is available not only for telemetry in filters, but for out-of-the-box SK telemetry as well, so I wanted to avoid confusion. But since you mentioned it here, I think it makes sense to add this to the example as well. Added scoping with transaction ID.

**Review by markwallace-microsoft — APPROVED** [review]

**Review by westey-m — APPROVED** [review]

File: dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    {
+        public async Task OnFunctionInvocationAsync(FunctionInvocationContext context, Func<FunctionInvocationContext, Task> next)
+        {
+            long startingTimestamp = Stopwatch.GetTimestamp();
+
+            logger.LogInformation("Function {FunctionName} invoking.", context.Function.Name);
+
+            if (context.Arguments.Count > 0)
+            {
+                logger.LogTrace("Function arguments: {Arguments}", JsonSerializer.Serialize(context.Arguments));
```
Comment (TaoChenOSU) [inline dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:117]: How does this work with the existing telemetry?

Comment (dmytrostruk) [reply dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:117]: It should not work with existing telemetry. This example contains the same telemetry which currently exists in SK, but recreated with Filters:
https://github.com/microsoft/semantic-kernel/blob/c596d25e270cc1caa906b6b3128eae684e55ac1d/dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs#L13-L16

As per original issue: https://github.com/microsoft/semantic-kernel/issues/5436
image

Comment (TaoChenOSU) [reply dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:117]: If we just want the logs, I think that's fine. But if we want the full telemetry, i.e. traces, and metrics, etc., there is still some difference between the filter approach and the out-of-the-box telemetry.

Comment (dmytrostruk) [reply dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs:117]: Makes sense. That's the reason why I explicitly mentioned logging here:
https://github.com/microsoft/semantic-kernel/blob/c596d25e270cc1caa906b6b3128eae684e55ac1d/dotnet/samples/Concepts/Filtering/TelemetryWithFilters.cs#L20-L21

If we will need more examples for traces and metrics we can add it later.
